pilot/pkg/networking/plugin/authz/rbac.go

```diff
@@ -0,0 +1,319 @@
+	namespace := split[1]
+
+	roles, err := store.List(model.ServiceRole.Type, namespace)
+	if err != nil {
+		log.Errorf("failed to get ServiceRoles in namespace %s: %v", namespace, err)
+		return nil
+	}
+
+	bindings, err := store.List(model.ServiceRoleBinding.Type, namespace)
+	if err != nil {
+		log.Errorf("failed to get ServiceRoleBinding in namespace %s: %v", namespace, err)
+		return nil
+	}
+
+	config, err := convertRbacRulesToFilterConfig(serviceName, roles, bindings)
+	if err != nil {
+		log.Errorf("failed to convert RBAC rules to filter config: %v", err)
+		return nil
+	}
+
+	log.Debugf("%s generated with config %v", RbacFilterName, *config)
+	return &http_conn.HttpFilter{
+		Name:   RbacFilterName,
+		Config: util.MessageToStruct(config),
+	}
+}
+
+func convertRbacRulesToFilterConfig(service string, roles []model.Config, bindings []model.Config) (*policy.RBAC, error) {
+	// roleToBinding maps ServiceRole to a list of ServiceRoleBindings.
+	roleToBinding := map[string][]*rbacproto.ServiceRoleBinding{}
+	for _, binding := range bindings {
+		proto := binding.Spec.(*rbacproto.ServiceRoleBinding)
+		refName := proto.RoleRef.Name
+		if refName == "" {
+			log.Errorf("invalid RoleRef.Name in binding: %v", proto)
+			continue
+		}
+		roleToBinding[refName] = append(roleToBinding[refName], proto)
+	}
+
+	rbac := &policy.RBAC{
+		// TODO(yangminzhu): Supports RBAC_DENY based on RbacConfig.
+		Action:   policy.RBAC_ALLOW,
+		Policies: map[string]*policy.Policy{},
+	}
+
+	for _, role := range roles {
+		// Constructs the policy for each ServiceRole.
+		policy := &policy.Policy{
+			Permissions: []*policy.Permission{},
+			Principals:  []*policy.Principal{},
+		}
+
+		principals := convertToPrincipals(roleToBinding[role.Name])
+		proto := role.Spec.(*rbacproto.ServiceRole)
+		hasPolicy := false
+		for _, rule := range proto.Rules {
+			if stringMatch(service, rule.Services) {
+				// Generates the policy if the service is matched to the services specified in ServiceRole.
+				policy.Permissions = append(policy.Permissions, convertToPermission(rule))
+				policy.Principals = principals
+				hasPolicy = true
+				log.Debugf("service %v matched role %v, generated policy %v", service, role.Name, *policy)
+			}
+		}
+
+		if hasPolicy {
+			rbac.Policies[role.Name] = policy
+		}
+	}
+
+	return rbac, nil
+}
+
+// convertToPermission converts a single AccessRule to a Permission.
+func convertToPermission(rule *rbacproto.AccessRule) *policy.Permission {
+	permission := &policy.Permission{}
+
+	if len(rule.Methods) > 0 {
+		permission.Methods = make([]string, len(rule.Methods))
+		copy(permission.Methods, rule.Methods)
+	}
+
+	if len(rule.Paths) > 0 {
+		permission.Paths = make([]*envoy_type.StringMatch, 0)
+		for _, path := range rule.Paths {
+			permission.Paths = append(permission.Paths, convertToStringMatch(path))
+		}
+	}
+
+	if len(rule.Constraints) > 0 {
+		conditions := make([]*policy.Permission_Condition, 0)
+		for _, v := range rule.Constraints {
```

Comment:
Same here. We have to check the "key" of the constraint first. We cannot put all the constraints to "header".

Reply:
Updated accordingly, see my reply above.